server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -448,6 +497,7 @@ public String getSourceAsString() {
      * The source of the document as a map (can be {@code null}).
      */
     public Map<String, Object> getSourceAsMap() {
+        assert hasReferences();
```

Comment:
Technically yes, but I went for being strict here. We really shouldn't have code touching released instances ever. Much easier to trust tests if we verify that fact in more rather than fewer spots IMO